lib/phoenix/pubsub/redis_server.ex

```diff
@@ -0,0 +1,164 @@
+    GenServer.start_link(__MODULE__, opts, name: Dict.fetch!(opts, :name))
+  end
+
+  @doc """
+  Initializes the server.
+
+  An initial connection establishment loop is entered. Once `:eredis_sub`
+  is started successfully, it handles reconnections automatically, so we
+  pass off reconnection handling once we find an initial connection.
+  """
+  def init(opts) do
+    {:ok, local_pid} = Phoenix.PubSub.Local.start_link()
+    opts = Dict.merge(@defaults, opts)
+    opts = Dict.merge(opts, host: String.to_char_list(to_string(opts[:host])),
+                            password: String.to_char_list(to_string(opts[:password])))
+
+    Process.flag(:trap_exit, true)
+    send(self, :establish_conn)
+
+    {:ok, struct(RedisServer, opts: opts,
+                              local_pid: local_pid,
+                              node_ref: :erlang.make_ref)}
+  end
+
+
+
+  def handle_call({:subscribe, pid, topic}, _from, state) do
+    {:reply, GenServer.call(state.local_pid, {:subscribe, pid, topic}), state}
+  end
+
+  def handle_call({:unsubscribe, pid, topic}, _from, state) do
+    {:reply, GenServer.call(state.local_pid, {:unsubscribe, pid, topic}), state}
+  end
+
+  def handle_call({:subscribers, topic}, _from, state) do
+    {:reply, GenServer.call(state.local_pid, {:subscribers, topic}), state}
+  end
+
+  def handle_call(:list, _from, state) do
+    {:reply, GenServer.call(state.local_pid, :list), state}
+  end
+
+  def handle_call({:broadcast, from_pid, topic, msg}, _from, state) do
+    with_conn state, fn state ->
+      result = :poolboy.transaction :phx_redis_pool, fn worker_pid ->
+        GenServer.call(worker_pid, {:publish_to_redis, "phx:#{topic}",
+                                   {1, state.node_ref, from_pid, msg}})
+      end
+      {:reply, result, state}
+    end
+  end
+
+  def handle_info({:pmessage, "phx:*", "phx:" <> topic, binary_msg, _client_pid}, state) do
+    :poolboy.transaction :phx_redis_pool, fn worker_pid ->
+      GenServer.cast(worker_pid, {:forward_to_subscribers,
+                                  state.local_pid, state.node_ref, topic, binary_msg})
+    end
+
+    :eredis_sub.ack_message(state.eredis_sub_pid)
+    {:noreply, state}
+  end
+
+  def handle_info({:EXIT, _pid, {:connection_error, {:connection_error, :econnrefused}}}, state) do
+    {:noreply, state}
+  end
+
+  @doc """
+  Connection establishment and shutdown loop
+
+  On init, an initial conection to redis is attempted when starting `:eredis_sub`.
+  If failed, the connection is tried again in `@reconnect_after_ms` until a max
+  of `@max_connect_attemps` is tried, at which point the server terminates with
+  an `:exceeded_max_conn_attempts`.
+  """
+  def handle_info(:establish_conn, %RedisServer{status: :connected} = state) do
+    {:noreply, state}
+  end
+  def handle_info(:establish_conn, %RedisServer{reconnect_attemps: count} = state)
+    when count >= @max_connect_attemps do
+
+    {:stop, :exceeded_max_conn_attempts, state}
```

Comment:
Let me start with some background since I am copying some folks here. This module is responsible to connect to Redis pubsub and receive notifications. Once we receive those notifications from Redis, we send them to the local subscribers. Currently we try to connect to Redis 3 times right after start, with 5 seconds interval, and then crash if we can't connect.

After thinking about this, I don't think we should crash. We should just continue trying to connect every X seconds. The reason is this: do you want the whole website to fail if we can't connect to Redis? Because that's what will happen after enough time. I would rather have that part of the website disabled.

If we do this change, we need to decide what is going to happen when someone tries to subscribe locally and there is no Redis connection. We have two options then:

1. We don't allow subscribers to connect in the first place. We change the subscribe function to return either `:ok` or `{:error, _}` and we will return an error if Redis is not available. The channel/transport/whatever should handle this case and not allow new connections.

2. We allow subscribers to connect but, because Redis is down, they won't be able to send any messages. We need to decide if being unable to send messages is a hard crash of the channel (and therefore the client) or if we should just return an `{:error, _}` tuple that will be handled in the channel (and hopefully notified to the client).

I am more inclined towards 2. That's because we don't need subscriptions to work to use channels. I may have a channel that only uses replies and never broadcasts.

The issue though is that currently developers are not handling the output of broadcast in their channels. Furthermore, handling the output for every broadcast will quickly be painful. Maybe we should just exit with something meaningful (as call and cast usually do) and allow the transport to handle those specially, often notifying the client. We could even provide a callback in JS that would be invoked for exactly the cases where you can't broadcast messages.

Btw, here is a great article on the topic: http://ferd.ca/it-s-about-the-guarantees.html

/cc @ericmj @fishcakez